feature/selector.py

```diff
@@ -270,6 +273,102 @@ def _validate(self):
+            * If num_feature is None, num_features defines by solving a set cover problem using unicost or diverse
+            The algorithm might choose a less number of features depending on other parameters.
+
+        featurization_method :
+            TextWiser object to featurize items in X_textual (it is named data in test and this method does not
+            have impact on the following options:
+                - Random optimization with num_features = max_num_features
+                - Greedy optimization with unicost as cost_metric
+
+        optimization_method : str, optional
+            * random: This method finds a random solution by selecting a random set of items.
+                           The number of selected times (num_features = t):
+                                - t = maximum number of features defines by user.
+                                      The cost_metric input argument needs to ignore in this case.
+                                - t = None: the number of feature computed by solving a set cover problem with
+                                      cost metrics (unicost or diverse)
+            * exact: This method finds the optimum solution using the Python-MIP package to solve a set cover.
+            * greedy: This method finds a greedy solution by adding items step by step
+                      using a greedy heuristic that covers the most labels.
+                      The text featurization uses to implement diverse cost metric.
+            * kmeans: This method clusters the text featurization space into k clusters
+                      where k is either the solution of the exact unicost selection, or,
+                      the given num_features. Then, items close the centroids are selected
+                      kmeans does not need cost metrics.
+           * max_cover: This method find the solution based on the multi-level optimization in the paper [1].
+                        num_feature must be an integer for third optimization level (solving max cover).
+
+        cost_metric : str, optional;
+            * unicost: Each item/feature incurs a cost of one when included in the selection.
+                        For random this parameter does not impact.
+            * diverse: Each item/feature incurs a cost equivalent to the distance to its closest centroid
+                       in latent space obtained from the text featurization of items.
+                       The centroids are found by clustering the text featurization space into k-clusters,
+                       where k is either the solution of the exact unicost selection, or,
+                       the given num_features.
+        """
+        # Defualt values
+        num_features: Union[int, None]
+        featurization_method: TextWiser = TextWiser(Embedding.TfIdf(min_df=10), Transformation.NMF(n_components=30))
+        optimization_method: str = "random"
```

Comment:
let's make the default as "exact" + "diverse" --which corresponds to the paper

Reply:
Then we need to add cost_metric=None for tests without a cost metric requirement.

Reply:
Not sure how the default behavior is related to None cost metric. 

Let's please not add a None to cost. Instead; 

For methods that don't depend on cost (say random()), we will document that clearly in the pydoc. 
Then, in the tests, we will write a test with each cost and verify that the results are the same (since the method is not dependant on cost)

Reply:
The default value of "None" for the cost metric has been removed. Additionally, an "if" condition has been added to the "_process_df" function to handle a scenario where the optimization method is "random", and we have a cost metric (either "unicost" or "diverse"), and the user has specified the number of features. In this case, then the text featurization method will NOT be executed.

To ensure that the results are consistent, a test has been created using two instances of "TextBased" - one with "cost_metric" set to "unicost" and the other with "cost_metric" set to "diverse". Both instances have the same number of features specified. The test checks whether the resulting indices are the same using "np.allclose()". If the indices are the same, the test passes.